### PR TITLE
fix: register correct unwrapped schema

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

This commit fixes several issues and refactors a lot of the serde code around wrapping and unwrapping single values. This needs to be done before we can support key formats that use wrapping / unwrapping.

The main issues being fixed are:
  1. allow each format to define if it supported wrapping and/or unwrapping. (Not possible with current design)
  2. pass the correct wrapping / unwrapping flags are passed to key vs value formats when creating serde. (bug in code passes same SerdeOptions to key and value).
  3. register the correct wrapped / unwrapped schema with the SR. (bug in existing code meant registered format is always wrapped).

At the same time, the way wrapping / unwrapping was handled in the code wasn't great. Formats like `JSON` needed to be able to handle both wrapped and unwrapped schemas and values, depending on whether the user _explicitly_ set wrapping or unwrapping, vs the default behaviour of the format. This commit refactors the code such that the format will always be passed the a consistent schema and the set of serde features the format should use when creating the serde. This simplifies things and paves the way to user-define-serde.

### Reviewing notes:

There's some refactoring of how physical and persistence schemas are created and used:
- `SerdeOptions` can now split the global options into key and value specific `EnabledSerdeFeatures` where,
- `EnabledSerdeFeatures` is a set of `SerdeFeature`s that have been validated to ensure there are no clashing features (currently checking wrap and unwrap are not both set).
- `PhysicalSchema` is now really just a combination `LogicalSchema` and `SerdeOptions`.
- `PersistenceSchema` is still using the Connect schema internally for now, but that schema will _always_ be a `STRUCT` containing the key/value columns. The instance also tracks the key/value serde features.

How serde are created has been refactored:
- The `KsqlSerdeFactory` is no more. Instead, 
- everything is hidden behind the `Format` interface, which now has a `createSerde` method, rather than one that returns a `KSqlSerdeFactory`.  This is much cleaner and simple.
- The old implementations of `KsqlSerdeFactory`  are now just implementation details of specific formats, and are free to be changed as needed.
- `Format` deals in terms of `SerdeFeature`s, not `SerdeOptions`.
- `GenericRowSerde` and `GenericKeySerde` are no longer responsible for handling wrapping and unwrapping, they just pass down and expect `Struct`.   (Common code has moved to `GenericSerdeFactory`).
- This means `Format.createSerde` can return correctly typed `Serde<Struct>`, rather than some unknown maybe `Struct` maybe some primitive.
 - Each format is responsible for dealing with the features it supports.  
 - `KafkaFormat` and `DelimitedFormat` can now support unwrapping and don't need to _do_ anything to support it.
 - Only Connect formats support both wrapping and unwrapping, so the `ConnectFormat` handles this by extracting the single field's schema and wrapping the serde created by the sub-class with serde that handle the unwrapping.

### Testing done 

usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java [hunk 1]
```diff
@@ -65,52 +74,100 @@ public boolean supportsSchemaInference() {
...
-      final SerdeOptions serdeOptions
-  ) {
-    final boolean singleField = valueConnectSchema.fields().size() == 1;
+    final Class<?> targetType = SchemaConverters.connectToJavaTypeConverter()
+        .toJavaType(innerSchema);
 
-    final boolean unwrapSingle = serdeOptions.valueWrapping()
-        .map(option -> option == SerdeOption.UNWRAP_SINGLE_VALUES)
-        .orElse(false);
+    if (schema.features().enabled(SerdeFeature.UNWRAP_SINGLES)) {
```
Comment (agavra) [inline ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java:130]: it seems a little weird to encapsulate this above (`applySinglesUnwrapping`) but then check for it explicitly here

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json [hunk 2]
```diff
@@ -39,6 +39,19 @@
...
+      "statements": [
+        "CREATE STREAM INPUT WITH (kafka_topic='input', value_format='AvRo', wrap_single_value=false);",
+        "CREATE STREAM OUTPUT as SELECT * FROM input;"
+      ],
+      "topics": [
+        {"name": "input", "format": "AVRO", "schema": "int"},
+        {"name": "OUTPUT", "format": "AVRO", "schema": {"name": "blah", "type": "record", "fields": [{"name": "ROWVAL", "type": "int"}]}}
+      ],
+      "inputs": [{"topic": "input", "value": 4}],
+      "outputs": [{"topic": "OUTPUT", "value": {"ROWVAL": 4}}]
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json:53]: does the output intentionally not keep the serde options from the input? I would expect it to be not wrapped here as well

**Review by agavra — APPROVED** [review]
LGTM I reviewed this change with a somewhat relaxed lens (it's hard to dig into these refactors at the code level) but what I checked looks good to me and your description is an overall positive change! Would have been nice to split it up into a few smaller PRs though 😅

File: ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SchemaConverters.java [hunk 3]
```diff
@@ -123,6 +123,16 @@ private SchemaConverters() {
     Schema toConnectSchema(SqlType sqlType, String name, String doc);
   }
 
+  public interface ConnectToJavaTypeConverter {
+    /**
+     * Convert the supplied Connect {@code schema} to its corresponding SQL type.
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SchemaConverters.java:128]: done.

File: ksqldb-common/src/main/java/io/confluent/ksql/serde/EnabledSerdeFeatures.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Validated set of enabled features
+ *
+ * <p>Known to not have conflicting features enabled
+ */
+@Immutable
+public final class EnabledSerdeFeatures {
+
+  private final ImmutableSet<SerdeFeature> features;
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-common/src/main/java/io/confluent/ksql/serde/EnabledSerdeFeatures.java:36]: All will become clear once `SerdeOption.WRAP_SINGLE_KEYS` is added, which will map to `SerdeFeature.WRAP_SINGLE` the same as `SerdeOption.WRAP_SINGLE_VALUES` does.

File: ksqldb-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java [hunk 5]
```diff
@@ -160,18 +162,16 @@ private static void produceOne(
   private Serializer<Struct> getKeySerializer(
       final ConnectSchema keySchema
   ) {
-    final PersistenceSchema schema = PersistenceSchema.from(
-        keySchema,
-        keySerializerFactory.format().supportedFeatures().contains(SerdeFeature.UNWRAP_SINGLES)
-    );
-
+    final Set<SerdeFeature> supported = keySerializerFactory.format().supportedFeatures();
+    final EnabledSerdeFeatures features = supported.contains(SerdeFeature.UNWRAP_SINGLES)
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java:166]: The code _wants_ to turn on unwrapping, where its supported, which is what it does. The code doesn't want to turn on any other feature the format may support, e.g. wrapping, or any feature added in the future.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroSchemas.java [hunk 6]
```diff
@@ -86,7 +86,12 @@ private static Schema buildAvroCompatibleSchema(
           return schema;
         }
 
-        schemaBuilder = new SchemaBuilder(schema.type());
+        schemaBuilder = new SchemaBuilder(schema.type())
+            .name(schema.name());
+
+        if (schema.parameters() != null) {
+          schemaBuilder.parameters(schema.parameters());
+        }
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroSchemas.java:94]: I'll take a look.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java:130]: Why?  `applySinglesUnwrapping(outerSchema)` only extracts the inner schema. That's only half the storey. The code still needs to build a serde that can handle the unwrapping, i.e. extracting the value of the single column from the `Struct` passed to `serialize` and the reverse for `deserialize`.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json:53]: `wrap_single_value` is not currently inherited.  Not sure why it was implemented that way, but it was.  This PR doesn't change that.

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroSchemas.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/avro/AvroSchemas.java]: done.

**Comment by big-andy-coates** [comment]
Thanks for the review @agavra.  Yeah, it was a big one.  Would of been a lot of work to break it into separate PRs and would have take a long time to chain the reviews together.  It was definitely a one of those times when you pull on a thread and the thing just keeps unraveling...

File: ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java (hunk 1 above)
Comment (agavra) [reply ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java:130]: I feel like if I'm already checking the `UNWRAP_SINGLES` variable here, i might as well just extract the inner schema there as well. Not a biggie, but now we're checking it in two places and doing a no-op in one of them if it's disabled.

File: ksqldb-common/src/main/java/io/confluent/ksql/serde/EnabledSerdeFeatures.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-common/src/main/java/io/confluent/ksql/serde/EnabledSerdeFeatures.java:36]: 💡

**Pull request merged**